packages/app/src/systems/Transaction/utils/tx.ts

```diff
@@ -430,7 +460,71 @@ export function getContractCallOperations({
         const newCallOps = contractCallReceipts.reduce(
           (prevContractCallOps, receipt) => {
             if (receipt.to === contractInput.contractID) {
-              const newContractCallOps = addOperation(prevOutputCallOps, {
+              const calls = [];
+
+              const abi = abiMap?.[contractInput.contractID];
+              if (abi) {
+                const abiInterface = new Interface(abi);
+                const callFunctionSelector = receipt.param1.toHex(8);
+                const functionFragment =
+                  abiInterface.getFunction(callFunctionSelector);
+
+                const nonEmptyInputs = filterEmptyParams(
+                  functionFragment.inputs
+                );
+
+                let encodedArgs;
+                if (functionFragment.isInputDataPointer()) {
+                  if (rawPayload) {
+                    const argsOffset = bn(receipt.param2)
+                      .sub(VM_TX_MEMORY)
+                      .toNumber();
+
+                    // slice(2) to remove first 0x, then slice again to remove offset and get only args
+                    encodedArgs = `0x${rawPayload
+                      .slice(2)
+                      .slice(argsOffset * 2)}`;
+                  }
+                } else {
+                  encodedArgs = receipt.param2.toHex();
+                }
+
+                let argumentsProvided;
+                if (encodedArgs) {
+                  const data = functionFragment.decodeArguments(encodedArgs);
+                  if (data) {
+                    argumentsProvided = nonEmptyInputs.reduce(
+                      (prev, input, index) => {
+                        const value = data[index];
+                        const name = input.name;
+
+                        if (name) {
+                          return {
+                            ...prev,
+                            // reparse to remove bn
+                            [name]: JSON.parse(JSON.stringify(value)),
+                          };
+                        }
+
+                        return prev;
+                      },
+                      {}
+                    );
+                  }
+                }
+
+                const call = {
+                  functionSignature: functionFragment.getSignature(),
+                  functionName: functionFragment.name,
+                  argumentsProvided,
+                  ...(receipt.amount?.isZero()
+                    ? {}
+                    : { amount: receipt.amount, assetId: receipt.assetId }),
+                };
+                calls.push(call);
+              }
```

Comment:
It's possible to split better this code and add some comments by instruction to make it better understandable for others?